tests/test_coaddBase.py

```diff
@@ -0,0 +1,134 @@
+        # a "jointcal_photoCalib" calibration to return
+        self.jointcalPhotoCalib = lsst.afw.image.PhotoCalib(1e-6, 1e-8)
+
+        crpix = lsst.geom.Point2D(0, 0)
+        crval = lsst.geom.SpherePoint(0, 45, lsst.geom.degrees)
+        cdMatrix = lsst.afw.geom.makeCdMatrix(scale=1.0*lsst.geom.arcseconds)
+        self.skyWcs = lsst.afw.geom.makeSkyWcs(crpix, crval, cdMatrix)
+        jointcalCdMatrix = lsst.afw.geom.makeCdMatrix(scale=0.9*lsst.geom.arcseconds)
+        # a "jointcal_wcs" skyWcs to return
+        self.jointcalSkyWcs = lsst.afw.geom.makeSkyWcs(crpix, crval, jointcalCdMatrix)
+
+        self.exposure = lsst.afw.image.ExposureF(10, 10)
+        self.exposure.maskedImage.image.array = np.random.random((10, 10)).astype(np.float32) * 1000
+        self.exposure.maskedImage.variance.array = np.random.random((10, 10)).astype(np.float32)
+        # mask at least one pixel
+        self.exposure.maskedImage.mask[5, 5] = 3
+        # set the Calib and Wcs objects of this exposure.
+        self.exposure.getCalib().setFluxMag0(1/meanCalibration, calibrationErr/meanCalibration**2)
+        self.exposure.setWcs(self.skyWcs)
+
+        # set to False in a test to return self.jointcalPhotoCalib for get('jointcal_PhotoCalib')
+        self.raiseOnGetPhotoCalib = True
+        # set to False in a test to return self.jointcalWcs for get('jointcal_Wcs')
+        self.raiseOnGetWcs = True
+
+        def mockGet(datasetType, dataId=None, immediate=None):
+            """Minimally fake a butler.get()."""
+            if "calexp" in datasetType:
+                return self.exposure.clone()
+            if "jointcal_photoCalib" in datasetType:
+                if self.raiseOnGetPhotoCalib:
+                    raise lsst.daf.persistence.NoResults("failed!", datasetType, dataId)
+                else:
+                    return self.jointcalPhotoCalib
+            if "jointcal_wcs" in datasetType:
+                if self.raiseOnGetWcs:
+                    raise lsst.daf.persistence.NoResults("failed!", datasetType, dataId)
+                else:
+                    return self.jointcalSkyWcs
+
+        self.dataRef = unittest.mock.Mock(spec=lsst.daf.persistence.ButlerDataRef)
+        self.dataRef.get.side_effect = mockGet
+        self.dataRef.dataId = {"ccd": 10000, "visit": 1}
+
+    def test_getCalibratedExposure(self):
+        expect = self.exposurePhotoCalib.calibrateImage(self.exposure.maskedImage)
+        expect /= self.exposurePhotoCalib.getCalibrationMean()
+        result = self.task.getCalibratedExposure(self.dataRef, True)
+
+        self.assertMaskedImagesEqual(result.maskedImage, expect)
+        # TODO: once RFC-545 is implemented, this should be 1.0
+        self.assertEqual(result.getCalib().getFluxMag0()[0], 1/self.exposurePhotoCalib.getCalibrationMean())
+        self.assertEqual(result.getWcs(), self.skyWcs)
+
+    def test_getCalibratedExposureNoJointcalPhotoCalib(self):
+        self.config.doApplyJointcal = True
```

Comment:
I think this is actually ok, because `setUp` runs (and creates a new `CoaddBaseConfig`, resetting state) before each test method is run.

Reply:
That's right. OK for the unit test behavior.  (That also means you don't save any time by doing it this way: changing the configs after creating the task.)

Changing a live task's configs by modifying a view into the config is still non-standard.  I'd be OK with it if you get @r-owen's blessing.  (At least add an inline comment deterring any onlookers from copying your pattern, and have `setup` set `doApplyUbercal=True` so you only do it once in `test_getCalibratedExposure`)

Reply:
When and how is the config frozen? Does it actually become *immutable*? Where is this documented? What do you mean by "a view into the config"?

I and others have used this trick in tests before: I wasn't aware there was anything odd about it.

What do you mean "don't save any time"? I don't do this to save time, I do it to reduce code duplication: create the config in `setUp`, modify it for individual tests as needed.

Reply:
To answer some of my own questions:

* `When and how is the config frozen?`: inside `ArgumentParser.parse_args()` by calling `config.freeze()`.
* `Does it actually become immutable?`: Yes, in that after `freeze()`, attempting to modify a config Field raises `FieldValidationError`.
* `Where is this documented?`: The Returns section of `parse_args` says "validated and frozen" for the `config` element of the returned argparse namespace. That's the only mention I could find, and I'd never have thought to look there for it.

All of the above suggests to me that there's nothing wrong with modifying a config if you never have an ArgumentParser, as in this case.

Reply:
Many tasks have constructor logic that depends on configuration (for instance, those that produce catalogs finalize their schemas after construction).  Allowing configuration to be modified after task construction for those could lead to hard-to-find errors logic bugs downstream (e.g. because the config and the schema are out of sync).  As a result, I've always advocated that *all* Task configuration be considered frozen after initialization, and in fact until recently I thought we enforced it in all Tasks.  It turns out we only enforce that in CmdLineTasks, but there's nothing that would make them special in this respect.

I've just created [DM-16548](https://jira.lsstcorp.org/browse/DM-16548) to fix that (I should have done that last week when I learned of the problem; glad this discussion reminded me of it).  Comments there on where to better document this are very welcome.

Once that lands, I think this test code will have to change, and if it's not too hard it might be best to change it up front - I can certainly believe that the local logic here is perfectly fine, but I think it's something of an anti-pattern that we don't want cargo-culted to other tests and tasks.

Reply:
I agree with @TallJimbo that tasks should always assume their configuration is frozen. For the record: it is the command parser that freezes the config (somebody has to do it). So the config received by the command line task and all its subtasks is frozen, and tasks should be written assuming that their config is frozen.

Reply:
For what it's worth, this technique is used in other tests in this package and (I suspect) elsewhere, some of them of quite long standing.; like @parejkoj, I was surprised to see it being called out as an anti-pattern.

The nearest we get to advising against this in official documentation is (as far as I've found) in the [pipe_base documentation](https://pipelines.lsst.io/v/d_2018_11_20/modules/lsst.pipe.base/creating-a-task.html#configuration) which asserts:

> The configuration is read-only; once you construct a task, the same configuration will be used to process all data with that task

From that wording, it's not clear if that's a statement of technical fact or of policy, though.

I'd be happy to see an update to that documentation — or, as DM-16548 suggests, a technical fix — to establish policy here.

Reply:
As I said above, the freezing happens inside `ArgumentParser`, so it is decoupled from the Task system (you don't need an `ArgumentParser` to write `Task` tests), and thus surprising to anyone not involved in developing `pipe_base`. I can rework these tests, but as @jdswinbank says, making config freezing happen in `Task` in DM-16548 will likely require changing quite a few tests.

I don't know if that requires an RFC: it *is* in interface change, but it's a change to what the system was apparently intended to be. (I won't argue against such a change: I'm just pointing out the distinction between the actual behavior here vs. what the original authors apparently intended.)